x/rollapp/transfergenesis/ibc_module.go

```diff
@@ -0,0 +1,258 @@
+	}
+
+	memo, err := getMemo(transfer.GetMemo())
+	if errorsmod.IsOf(err, gerr.ErrNotFound) {
+		l.Debug("Memo not found.")
+		// If someone tries to send a transfer without the memo before the bridge is open, they will
+		// be blocked at the transfersenabled middleware
+		return w.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+	if err != nil {
+		l.Error("Get memo.", "err", err)
+		return channeltypes.NewErrorAcknowledgement(errorsmod.Wrap(err, "get memo"))
+	}
+
+	ra := transfer.Rollapp
+
+	nTransfersDone, err := w.rollappKeeper.VerifyAndRecordGenesisTransfer(ctx, ra.RollappId, memo.ThisTransferIx, memo.TotalNumTransfers)
+	if errorsmod.IsOf(err, derr.ErrViolatesDymensionRollappStandard) {
+		// The rollapp has deviated from the protocol!
+		handleFraudErr := w.handleFraud(ra.RollappId)
+		if err != nil {
+			l.Error("Handling fraud.", "err", handleFraudErr)
+		} else {
+			l.Info("Handled fraud: verify and record genesis transfer.", "err", err)
+		}
+		return channeltypes.NewErrorAcknowledgement(errorsmod.Wrap(err, "transfer genesis"))
+	}
+	if err != nil {
+		l.Error("Verify and record transfer.", "err", err)
+		return channeltypes.NewErrorAcknowledgement(errorsmod.Wrap(err, "transfer genesis: verify and record"))
+	}
+
+	// it's a valid genesis transfer!
+
+	err = w.registerDenomMetadata(ctx, ra.RollappId, ra.ChannelId, memo.Denom)
+	if err != nil {
+		l.Error("Register denom metadata.", "err", err)
+		return channeltypes.NewErrorAcknowledgement(errorsmod.Wrap(err, "transfer genesis: register denom metadata"))
+	}
+
+	l.Debug("Received valid genesis transfer. Registered denom data.",
+		"num total", memo.TotalNumTransfers,
+		"num received so far", nTransfersDone,
+		"this transfer ix", memo.ThisTransferIx,
+	)
+
+	if nTransfersDone == memo.TotalNumTransfers {
+		// The transfer window is finished! Queue up a finalization
+		ctx.EventManager().EmitEvent(allTransfersReceivedEvent(ra.RollappId, nTransfersDone))
+		w.rollappKeeper.EnableTransfers(ctx, ra.RollappId)
+		l.Info("All genesis transfers received, bridge opened.",
+			"rollapp", ra.RollappId,
+			"n transfers", nTransfersDone)
+	}
+
+	l.Debug("Passing on the transfer down the stack, but skipping delayedack, the transferEnabled blocker and bridging fee.")
+
+	return w.IBCModule.OnRecvPacket(
+		bridgingfee.SkipContext(
+			transfersenabled.SkipContext(
+				delayedack.SkipContext(ctx),
+			),
+		),
+		packet, relayer)
+}
+
+// handleFraud : the rollapp has violated the DRS!
+func (w IBCModule) handleFraud(raID string) error {
+	// TODO: see https://github.com/dymensionxyz/dymension/issues/930
+	return nil
+}
+
+func allTransfersReceivedEvent(raID string, nReceived uint64) sdk.Event {
+	return sdk.NewEvent(types.EventTypeTransferGenesisAllReceived,
+		sdk.NewAttribute(types.AttributeKeyRollappId, raID),
+		sdk.NewAttribute(types.AttributeKeyTransferGenesisNReceived, strconv.FormatUint(nReceived, 10)),
+	)
+}
+
+func getMemo(rawMemo string) (rollapptypes.GenesisTransferMemo, error) {
```

Comment:
maybe call it `getMemoData` or `getGenTransferMemo`

Reply:
it's already in the transfer genesis module
so it's fine IMO